feat: Otel pytest plugin (#1217)
elastic/apm-pipeline-library: resources/scripts/pytest_otel/README.md

Features
--------

> pytest-otel plugin for report APM traces about test executed.

Reviewer: ```suggestion
pytest-otel plugin for reporting APM traces of tests executed.
```

Requirements
------------

* opentelemetry-api == 1.2
* opentelemetry-exporter-otlp == 1.2.0
* opentelemetry-sdk == 1.2.0

Installation
------------

You can install "pytest-otel" via `pip` or using the `setup.py` script.

```
git checkout https://github.com/elastic/apm-pipeline-library
cd apm-pipeline-library/resources/scripts
pip install ./pytest_otel
```

Usage
-----

> pytest_otel is configured by adding some parameters to the pytest command line here are the descriptions

Reviewer: ```suggestion
`pytest_otel` is configured by adding some parameters to the pytest command line. Below are the descriptions:
```

> * --endpoint: URL for the OpenTelemetry server. (Required). Env variable `OTEL_EXPORTER_OTLP_ENDPOINT`

Reviewer: ```suggestion
* --endpoint: URL for the OpenTelemetry server. (Required). Env variable: `OTEL_EXPORTER_OTLP_ENDPOINT`
```

> * --headers: Additional headers to send (i.e.: key1=value1,key2=value2). Env variable `OTEL_EXPORTER_OTLP_HEADERS`

Reviewer: ```suggestion
* --headers: Additional headers to send (i.e.: key1=value1,key2=value2). Env variable: `OTEL_EXPORTER_OTLP_HEADERS`
```

> * --service-name: Name of the service. Env variable `OTEL_SERVICE_NAME`

Reviewer: ```suggestion
* --service-name: Name of the service. Env variable: `OTEL_SERVICE_NAME`
```

> * --session-name: Name for the Main span reported.

Reviewer: ```suggestion
* --session-name: Name for the main span.
```

> * --trace_id: Trace parent id. Env variable `TRACE_ID` see https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format

Reviewer 2: Did you consider supporting consuming `traceparent` as an alternative to `trace_id`?

Author: the content is `traceparent` as it is specified at https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format. I do not really found documentation about `TRACE_ID` environment variable.

Reviewer 2: I'm wondering if consuming a `traceparent` rather than `trace_id` wouldn't be better because what we need to connect a span to a parent span is not only the parent trace_id but also the parent span_id and the `trace_flags` to get the sampling decision `sampled`.
Note that the [`otel_cli`](https://github.com/packethost/otel-cli) typically consumes the `TRACEPARENT` environment header

Author: I guess I have to rename the flag because is not acurate

Author: ```suggestion
* --traceparent: Trace parent id. Env variable `TRACEPARENT` see https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format
```

> * --traceparent: Trace parent id. Env variable `TRACEPARENT` see https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format

Reviewer: ```suggestion
* --traceparent: Trace parent ID. Env variable: `TRACEPARENT`. See https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format
```

> * --insecure: Disables TLS. Env variable `OTEL_EXPORTER_OTLP_INSECURE`

Reviewer: ```suggestion
* --insecure: Disables TLS. Env variable: `OTEL_EXPORTER_OTLP_INSECURE`
```

```bash
cd pytest_otel

> pytest --endpoint https://apm.example.com:8200 \

Reviewer 2: ```suggestion
pytest --endpoint https://otelcollector.example.com:4317 \
```

       --headers "Authorization=Bearer ASWDCcCRFfr" \
       --service-name pytest_otel \
       --session-name='My_Test_cases' \

>        --trace_id=00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01 \

Reviewer 2: The pattern of this value looks like a `traceparent` more than a `trace_id`

Author: it is a traceparent, I did not found documentation or samples about the content of `TRACE_ID` environment variable so I guessed that it is the content of the header `traceparent` as it is specify by https://www.w3.org/TR/trace-context-1/#trace-context-http-headers-format

Author: ```suggestion
       --traceparent=00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01 \
```

       --insecure=False
```

**IMPORTANT**: If you use `--headers` the transaction metadata might expose those arguments
with their values. In order to avoid any credentials to be exposed, it's recommended to use the environment variables.

> For instance, given the above example, the similar one with environment variables can be seen below:

Reviewer: ```suggestion
For instance, given the above example, a similar one with environment variables can be seen below:
```

```bash
OTEL_EXPORTER_OTLP_ENDPOINT=https://apm.example.com:8200 \
OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer ASWDCcCRFfr" \
OTEL_SERVICE_NAME=pytest_otel \

> TRACE_ID=00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01 \

Author: ```suggestion
TRACEPARENT=00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01 \
```

> OTEL_EXPORTER_OTLP_INSECURE=False \

Reviewer 2: why do you need to specify insecure=false?

Author: by default is False, this is a command example with all settings in it as an example.
